src/Data/DataResponse.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+namespace Yiisoft\Yii\Web\Data;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A wrapper around PSR-7 response that is assigned raw data to be formatted later using a formatter.
+ *
+ * For example, `['name' => 'Dmitriy']` to be formatted to JSON using {@see \Yiisoft\Yii\Web\Data\Formatter\JsonDataResponseFormatter}
+ * when {@see DataResponse::getBody()} is called.
+ */
+class DataResponse implements ResponseInterface
+{
+    private ResponseInterface $response;
+
+    private $data;
+
+    private ?StreamInterface $dataStream = null;
+
+    private ?DataResponseFormatterInterface $responseFormatter = null;
+
+    public function __construct($data, int $code, string $reasonPhrase, ResponseFactoryInterface $responseFactory)
+    {
+        $this->response = $responseFactory->createResponse($code, $reasonPhrase);
+        $this->data = $data;
+    }
+
+    public function getBody(): StreamInterface
+    {
+        if ($this->dataStream !== null) {
+            return $this->dataStream;
+        }
+
+        if ($this->data === null) {
+            return $this->dataStream = $this->response->getBody();
+        }
+
+        if ($this->responseFormatter !== null) {
+            $this->response = $this->responseFormatter->format($this);
+            return $this->dataStream = $this->response->getBody();
+        }
+
+        $data = $this->getData();
+        if (is_string($data)) {
+            $this->response->getBody()->write($data);
+            return $this->dataStream = $this->response->getBody();
+        }
+
+        throw new \RuntimeException('Data must be a string value.');
+    }
+
+    public function getHeader($name): array
+    {
+        return $this->response->getHeader($name);
+    }
+
+    public function getHeaderLine($name): string
+    {
+        return $this->response->getHeaderLine($name);
+    }
+
+    public function getHeaders(): array
+    {
+        return $this->response->getHeaders();
+    }
+
+    public function getProtocolVersion(): string
+    {
+        return $this->response->getProtocolVersion();
+    }
+
+    public function getReasonPhrase(): string
+    {
+        return $this->response->getReasonPhrase();
+    }
+
+    public function getStatusCode(): int
+    {
+        return $this->response->getStatusCode();
+    }
+
+    public function hasHeader($name): bool
+    {
+        return $this->response->hasHeader($name);
+    }
+
+    public function withAddedHeader($name, $value): DataResponse
```

Comment:
```suggestion
    public function withAddedHeader($name, $value): self
```
Please fix all places

Reply:
3d5bf42b1dab333cf439e960bc0009c874e79d07